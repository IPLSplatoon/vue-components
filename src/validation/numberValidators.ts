import { ValidatorResult } from './validator';

export function maxValue(maxValue: number): (value?: number | null) => ValidatorResult {
    return (value?: number | null) => {
        return {
            isValid: value === null || value === undefined || maxValue >= value,
            message: `Must not be above ${maxValue}`
        };
    };
}

export function minValue(minValue: number): (value?: number | null) => ValidatorResult {
    return (value?: number | null) => {
        return {
            isValid: value === null || value === undefined || value >= minValue,
            message: `Must not be below ${minValue}`
        };
    };
}
