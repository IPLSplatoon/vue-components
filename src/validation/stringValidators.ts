import { ValidatorResult } from './validator';
import { pluralize } from '../helpers/stringHelper';
import trim from 'lodash/trim';
import isEmpty from 'lodash/isEmpty';

export function minLength(length: number): (value?: string | null) => ValidatorResult {
    return (value?: string | null) => {
        return {
            isValid: isEmpty(value) || (value?.length ?? 0) >= length,
            message: `Must be at least ${pluralize('character', length)}`
        };
    };
}

export function maxLength(length: number): (value?: string | null) => ValidatorResult {
    return (value?: string | null) => {
        return {
            isValid: isEmpty(value) || length >= (value?.length ?? 0),
            message: `Must not be over ${pluralize('character', length)}`
        };
    };
}

export const numeric: (value?: string | null) => ValidatorResult = (value?: string | null) => ({
    isValid: isEmpty(value) || /^\d*$/.test(value ?? ''),
    message: 'Must be numeric'
});

export const notBlank: (value: string) => ValidatorResult = (value) => ({
    isValid: trim(value) !== '',
    message: 'Must not be blank'
});
