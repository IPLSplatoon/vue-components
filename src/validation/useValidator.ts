import { inject, InjectionKey, MaybeRefOrGetter, onUnmounted, toValue, watch } from 'vue';
import { FieldValidatorState, ValidatorResult } from './validator';

export const ValidatorInjectionKey = Symbol() as InjectionKey<{
    state: Record<string, FieldValidatorState<unknown>>,
    update: (name: string, isValid: boolean | null, message?: string | null) => void
}>;

export function useValidator<T>(
    name: MaybeRefOrGetter<string>,
    value: MaybeRefOrGetter<T>
): ValidatorResult {
    const validators = inject(ValidatorInjectionKey, undefined);
    const validator = validators?.state[toValue(name)];

    if (validator == null) {
        return {
            isValid: null,
            message: null
        };
    }

    watch(
        () => toValue(value),
        newValue => {
            for (let i = 0; i < validator.definition.validators.length; i++) {
                const validationResult = validator.definition.validators[i](newValue);
                if (!validationResult.isValid) {
                    validators?.update(toValue(name), false, validationResult.message);
                    return;
                }
            }

            validators?.update(toValue(name), true, null);
        },
        { immediate: validator.definition.immediate ?? false });

    onUnmounted(() => {
        validators?.update(toValue(name), null, null);
    });

    return validator.result;
}
