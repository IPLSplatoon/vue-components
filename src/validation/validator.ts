import { computed, ComputedRef, provide, reactive } from 'vue';
import { ValidatorInjectionKey } from './useValidator';

export interface ValidatorResult {
    isValid: boolean | null
    message?: string | null
}

export interface FieldValidatorDefinition<T> {
    immediate: boolean
    validators: ValidatorFunction<T>[]
}

export interface FieldValidatorState<T> {
    definition: FieldValidatorDefinition<T>
    result: ValidatorResult
}

export type ValidatorFunction<T> = ((value: T) => ValidatorResult)

export function validator<T>(
    immediate: boolean,
    ...validators: ValidatorFunction<T>[]
): FieldValidatorDefinition<T> {
    return {
        immediate,
        validators
    };
}

export function provideValidators(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    validators: Record<string, FieldValidatorDefinition<any>>
): { allValid: ComputedRef<boolean>, fieldIsValid: (name: string) => boolean } {
    const state = Object.entries(validators)
        .reduce((result: Record<string, FieldValidatorState<unknown>>, [name, definition]) => {
            result[name] = {
                definition,
                result: reactive({
                    isValid: null,
                    message: null
                })
            };
            return result;
        }, { });

    provide(ValidatorInjectionKey, {
        state,
        update(name, isValid, message) {
            if (state[name]) {
                state[name].result.isValid = isValid;
                state[name].result.message = message;
            }
        }
    });

    return {
        allValid: computed(() => Object.values(state).every(status => status.result.isValid)),
        fieldIsValid: name => state[name]?.result.isValid === true
    };
}
