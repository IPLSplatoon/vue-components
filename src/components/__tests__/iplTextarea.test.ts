import { mount } from '@vue/test-utils';
import IplTextarea from '../iplTextarea.vue';
import { ValidatorInjectionKey } from '../../validation/useValidator';

describe('iplTextarea', () => {
    it('matches snapshot', () => {
        const wrapper = mount(IplTextarea, {
            props: {
                modelValue: null,
                name: 'test-textarea',
                label: 'Textarea!',
                placeholder: 'Test Placeholder',
                disabled: false
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot when valid', () => {
        const wrapper = mount(IplTextarea, {
            props: {
                modelValue: null,
                name: 'test-textarea',
                label: 'Textarea!',
                placeholder: 'Test Placeholder',
                disabled: false
            },
            global: {
                provide: {
                    [ValidatorInjectionKey as symbol]: {
                        state: {
                            'test-textarea': {
                                definition: {
                                    immediate: false
                                },
                                result: {
                                    isValid: true,
                                    message: null
                                }
                            }
                        }
                    }
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot when invalid', () => {
        const wrapper = mount(IplTextarea, {
            props: {
                modelValue: null,
                name: 'test-textarea',
                label: 'Textarea!',
                placeholder: 'Test Placeholder',
                disabled: false
            },
            global: {
                provide: {
                    [ValidatorInjectionKey as symbol]: {
                        state: {
                            'test-textarea': {
                                definition: {
                                    immediate: false
                                },
                                result: {
                                    isValid: false,
                                    message: 'error!!'
                                }
                            }
                        }
                    }
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('emits event on input change', async () => {
        const wrapper = mount(IplTextarea, {
            props: {
                modelValue: null,
                name: 'test-textarea',
                label: 'Textarea!',
                placeholder: 'Test Placeholder',
                disabled: false
            }
        });

        const textarea = wrapper.get('textarea');
        textarea.element.value = 'new-value';
        await textarea.trigger('input');

        expect(wrapper.emitted('update:modelValue')).toEqual([['new-value']]);
    });
});
