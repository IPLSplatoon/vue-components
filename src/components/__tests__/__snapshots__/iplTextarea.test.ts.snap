// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iplTextarea matches snapshot 1`] = `
"<div class="ipl-textarea__wrapper"><label class="ipl-label">Textarea! <textarea name="test-textarea" placeholder="Test Placeholder"></textarea></label>
  <!--v-if-->
</div>"
`;

exports[`iplTextarea matches snapshot when invalid 1`] = `"<div class="ipl-textarea__wrapper has-error"><label class="ipl-label">Textarea! <textarea name="test-textarea" placeholder="Test Placeholder"></textarea></label><span class="error">error!!</span></div>"`;

exports[`iplTextarea matches snapshot when valid 1`] = `
"<div class="ipl-textarea__wrapper"><label class="ipl-label">Textarea! <textarea name="test-textarea" placeholder="Test Placeholder"></textarea></label>
  <!--v-if-->
</div>"
`;
